modules/webgl/src/context/context.js

```diff
@@ -200,27 +203,38 @@ function getVersion(gl) {
 // use devicePixelRatio to set canvas width and height
 function setDevicePixelRatio(gl, devicePixelRatio, options) {
   let devicePixelRatioClamped = false;
-  let aspectRatioValid = false;
+  let logWarning = false;
 
   // NOTE: if options.width and options.height not used remove in v8
   const clientWidth =
     'width' in options ? options.width : gl.canvas.clientWidth || gl.canvas.width || 1;
   const clientHeight =
     'height' in options ? options.height : gl.canvas.clientHeight || gl.canvas.height || 1;
 
-  // Note: when devicePixelRatio is too high, it is possible we might hit system limit for
-  // drawing buffer width and hight, in those cases they get clamped and resulting aspect ration may not be maintained
-  // for those cases, reduce devicePixelRatio.
-  do {
-    gl.canvas.width = Math.ceil(clientWidth * devicePixelRatio);
-    gl.canvas.height = Math.ceil(clientHeight * devicePixelRatio);
-    aspectRatioValid =
-      gl.drawingBufferWidth / clientWidth === gl.drawingBufferHeight / clientHeight;
-    devicePixelRatio = Math.max(devicePixelRatio / 2, 1);
-    devicePixelRatioClamped = devicePixelRatioClamped || !aspectRatioValid;
-  } while (!aspectRatioValid);
-
-  if (devicePixelRatioClamped) {
-    log.warn(`Device pixel ratio clamped`)();
+  const cachedSize = gl._canvasSizeInfo;
+  // Check if canvas needs to be resized
+  if (
+    cachedSize.clientWidth !== clientWidth ||
+    cachedSize.clientHeight !== clientHeight ||
+    cachedSize.devicePixelRatio !== devicePixelRatio
+  ) {
+    // Note: when devicePixelRatio is too high, it is possible we might hit system limit for
+    // drawing buffer width and hight, in those cases they get clamped and resulting aspect ration may not be maintained
+    // for those cases, reduce devicePixelRatio.
+    let clampedPixelRatio = devicePixelRatio;
+    do {
+      const canvasWidth = Math.floor(clientWidth * clampedPixelRatio);
+      const canvasHeight = Math.floor(clientHeight * clampedPixelRatio);
+      gl.canvas.width = canvasWidth;
+      gl.canvas.height = canvasHeight;
+      devicePixelRatioClamped =
+        gl.drawingBufferWidth !== canvasWidth || gl.drawingBufferHeight !== canvasHeight;
+      clampedPixelRatio = Math.max(clampedPixelRatio / 2, 1);
+      logWarning = logWarning || devicePixelRatioClamped;
+    } while (devicePixelRatioClamped);
+    if (logWarning) {
+      log.warn(`Device pixel ratio clamped`)();
```

Comment:
`log.warn` only logs the same warning once. You do not need the custom flag `logWarning`.